### PR TITLE
Coqdoc links

### LINKED ISSUE #13786: Coqdoc creates non-working links in the latex/pdf output
#### Description of the problem

The command below creates a pdf file with a html-link to `nat` in the documentation.
However, the code in tools/coqdoc/output.ml creates bad latex code, so the link
doesnt work.
 
```
echo 'Variable a:nat.' > Bar.v && coqc Bar.v && coqdoc --pdf Bar.v > Bar.pdf
```

#### Coq Version

version 8.13.0

### PR DESCRIPTION
**Pull request by larsr**
**Kind:** bug fix 

Fixes / closes #13786

- [x] Added / updated test-suite

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: tools/coqdoc/output.ml [hunk 1]
```diff
@@ -260,6 +260,40 @@ module Latex = struct
   let label_ident s =
     for i = 0 to String.length s - 1 do label_char s.[i] done
 
+
+  let convert_string f s =
+    let n = String.length s in
+    let rec qmap i =
+      if i == n then "" else (f s.[i]) ^ (qmap (i + 1))
+    in qmap 0
```
Comment (silene) [inline tools/coqdoc/output.ml:268]: I know that `coqdoc` is not critical performance-wise. But here that is pushing it a bit too far. Please use the `Buffer` module to build your string.

File: tools/coqdoc/output.ml [hunk 2]
```diff
@@ -260,6 +260,40 @@ module Latex = struct
...
+
+   let string_contains f s =
+     let n = String.length s in
+     let rec find i =
+        if i == n then false else if f s.[i] then true else find (i+1)
+      in find 0
+
+  let encode_fragment s =
+     if string_contains (not_fragment_char) s
+     then Digest.to_hex (Digest.string s)
```
Comment (silene) [inline tools/coqdoc/output.ml:286]: What I said in #13787 still holds. The fragment of the URL in the LaTeX backend has to be identical to the one in the HTML backend. (They point to the exact same object on the Internet.) So, please use the same function for both backends. Do not let the code give the impression they are unrelated.

File: tools/coqdoc/output.ml [hunk 3]
```diff
@@ -340,8 +374,8 @@ module Latex = struct
           printf "\\coqref{"; label_ident id;
           printf "}{\\coqdoc%s{%s}}" (type_name typ) s
       | External m when !externals ->
-          printf "\\coqexternalref{"; label_ident fid;
-          printf "}{%s}{\\coqdoc%s{%s}}" (escaped m) (type_name typ) s
+          printf "\\coqexternalref{%s}{%s}" (escaped m) (escaped (encode_fragment fid));
+          printf "{\\coqdoc%s{%s}}" (type_name typ) s
```
Comment (silene) [inline tools/coqdoc/output.ml:378]: Just use a single `printf`.

File: tools/coqdoc/output.ml [hunk 4]
```diff
@@ -260,6 +260,40 @@ module Latex = struct
...
+             | '<' | '>' | '&' | '\'' | '"' | '-' | ':' -> true
+        | _ -> false
+
+   let not_fragment_char c = not (fragment_char c)
+
+   let string_contains f s =
+     let n = String.length s in
+     let rec find i =
+        if i == n then false else if f s.[i] then true else find (i+1)
+      in find 0
```
Comment (silene) [inline tools/coqdoc/output.ml:282]: `f s.[i] || find (i+1)` might be a bit more readable.

File: tools/coqdoc/output.ml [hunk 5]
```diff
@@ -260,6 +260,40 @@ module Latex = struct
...
+
+  let encode_fragment s =
+     if string_contains (not_fragment_char) s
+     then Digest.to_hex (Digest.string s)
+     else
+       convert_string (fun c ->
+         if c >= '\x80' then Printf.sprintf "x%X" (int_of_char c) else
```
Comment (silene) [inline tools/coqdoc/output.ml:289]: Unless I am missing something, this case should already have been handled by the `Digest.string` case.

File: tools/coqdoc/output.ml [hunk 6]
```diff
@@ -260,6 +260,40 @@ module Latex = struct
...
+         match c with
+           | ' ' -> "_"
```
Comment (silene) [inline tools/coqdoc/output.ml:291]: Same here.

**Review by silene — CHANGES_REQUESTED** [review]

File: tools/coqdoc/output.ml (hunk 1 above)
Comment (larsr) [reply tools/coqdoc/output.ml:268]: Is it ok to use ocaml 4.07?  There is so much missing in 4.05, like String.to_seq and Seq.flat_map.

Comment (gares) [reply tools/coqdoc/output.ml:268]: Up to now no, we still declare 4.05 to be the minor version, but I'll raise the point this afternoon at the Coq Call.

Comment (silene) [reply tools/coqdoc/output.ml:268]: Honestly, `String.to_seq` is not worth the effort here. If you use `Buffer`, then you can use a plain `String.iter`.

**Comment by herbelin** [comment]
A remark in passing. To have a retro-active fix and preserve the compatibility of existing files produced by coqdoc (when escaping is not needed), why not to swap instead the `#1` and `#2` of `\coqexternalref` in `coqdoc.sty`?

**Comment by silene** [comment]
That would only help users who are mixing the `coqdoc` of an old Coq with the `coqdoc.sty` of a new Coq. Honestly, I do not think it is a good motivation, and we should ignore it. However, making the churn in the test-suite more palatable is a good motivation for changing the order in `coqdoc.sty` rather than `output.ml`.

**Comment by herbelin** [comment]
>  making the churn in the test-suite more palatable

Yes, that's a good argument. And for previous versions, we can also tell users who may like a fix for previous versions to hack their coqdoc.sty.

**Comment by larsr** [comment]
I have updated the code in another branch, coqdoc-links-2, but is it preferable to push to this branch so that the discussion and advice is not "lost"?
The -2 branch is rebased onto a newer coq/master to apply cleanly,so i'd like to not have to add unnecessary "merge" patches. But how can I do that without messing up this (coqdoc-links) branch?
[as a newbie] What is the preferred way of doing this?

**Comment by gares** [comment]
I think you should:
```shell
git branch old-attempt coqdoc-links # you keep a copy of this branch just in case
git checkout coqdoc-links
git reset --hard coqdoc-links-2
git push coqdoc-links -f
```

Comments will not be lost, the code snippet they apply to will be displayed as outdated.

File: tools/coqdoc/output.ml [hunk 7]
```diff
@@ -107,6 +107,50 @@ let set_module m sub = current_module := (m,sub);
...
+  | '>' | '<' | '#' | '&' as c -> sprintf "%%%X" (int_of_char c)
+  | ' ' -> "_"
+  | c -> String.make 1 c
+  )
+
+(* A 'fragment' is the part in the uri path that comes after the '#'. *)
+
+let allowed_fragment_char = function
+  | 'a'..'z' | 'A'..'Z' | '0'..'9' | '.' | '_' | '<' | '>' | '&'
+    | '\'' | '"' | '-' | ':' -> true
```
Comment (silene) [inline tools/coqdoc/output.ml:139]: This leads to the following (notice the `%3E` part).
```latex
\coqexternalref{http://coq.inria.fr/stdlib/Coq.Init.Logic}{::type\_scope:x\_'-\%3E'\_x}{\coqdocnotation{\ensuremath{\rightarrow}}}
```

Either you allow only pure fragment characters (and use `Digest.string` for nonconforming fragments) or you encode every forbidden character as `%XX` (and never use `Digest.string`). But please do not use a mix of both. (I have a small preference for the former, i.e., removing `<`, `>` from the list above.)

By the way, the list seems incomplete. I think `/`, `?`, `:`, `@`, `~`, `!`, `$`, `(`, `)`, `*`, `+`, `,`, `;`, `=` are also allowed. Where they omitted on purpose? (Backward compatibility?)

File: tools/coqdoc/output.ml [hunk 8]
```diff
@@ -107,6 +107,50 @@ let set_module m sub = current_module := (m,sub);
...
+  | _ -> false
+
+let uri_fragment_escape s =
+  let string_contains f s =
+     let rec find i = i < String.length s && (f s.[i] || find (i+1))
+     in find 0 in
+
+  if string_contains (not % allowed_fragment_char) s
```
Comment (silene) [inline tools/coqdoc/output.ml:147]: A bit too many indirections for my taste, I had a hard time reading it. Here is a suggestion:
```ocaml
let uri_fragment_escape s =
  let l = String.length s in
  let rec valid i = if i = l then true else allowed_fragment_char s.[i] && valid (i + 1) in
  if valid 0 then
```

File: tools/coqdoc/output.ml [hunk 9]
```diff
@@ -107,6 +107,50 @@ let set_module m sub = current_module := (m,sub);
...
+  then Digest.to_hex (Digest.string s)
+  else string_flatmap (function
+    | '>' | '<' | '#' | '&' as c -> sprintf "%%%X" (int_of_char c)
```
Comment (silene) [inline tools/coqdoc/output.ml:150]: `&` does not need to be encoded; it is a valid fragment character.

**Review by silene — CHANGES_REQUESTED** [review]

**Comment by SkySkimmer** [comment]
Feel free to reopen or open a new PR if there is progress.

**Pull request closed without merging**